Fix premature close of build output on pull

jlhawn opened the pull request:
The `build` job will sometimes trigger a `pull` job when the base image
does not exist. Now that engine jobs properly close their output by default
the pull job would also close the build job's stdout in a cascading close
upon completion of the pull.

This patch corrects this to set the `pull` job to not close it's IO when
complete. It also adds a warning log when a job writes to output with no
destinations and makes it an error to call `Write()` on a job output that
has already been closed.

fixes #10364

jlhawn commented:
Another fix would have been to make the `pull` job with stdout wrapped in an `ioutil.NopCloser`.

jlhawn commented:
er, the equivalent version of `ioutil.NopCloser` but for a writer.

dqminh commented:
I think wrapping `pull` job stdout is probably a cleaner fix.

dmcgowan commented on engine/streams.go:
```diff
@@ -82,9 +86,19 @@ func (o *Output) AddPipe() (io.Reader, error) {
 // Write writes the same data to all registered destinations.
 // This method is thread-safe.
 func (o *Output) Write(p []byte) (n int, err error) {
+	if o.closed {
```
Shouldn't this be done inside the lock to prevent a race condition

Commit:
Fix premature close of build output on pull

The build job will sometimes trigger a pull job when the base image
does not exist. Now that engine jobs properly close their output by default
the pull job would also close the build job's stdout in a cascading close
upon completion of the pull.

This patch corrects this by wrapping the `pull` job's stdout with a
nopCloseWriter which will not close the stdout of the `build` job.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

jlhawn commented:
While I think better close handling in the engine Output is also a good idea in the long run, I've updated this patch to just wrap the `pull` job's stdout with a NopCloseWriter for now as the first fix breaks too many tests.

tiborvass commented:
@jlhawn much better, and solves the problem!

LGTM

jlhawn commented:
@tiborvass said:

> much better

I disagree that it's better, but I'll fix the `engine.Output` stuff later.

unclejack commented:
@jlhawn Can you also add a test for this, please?

jlhawn commented:
@unclejack I added a unit test to the `engine` package with a note to fix the write-on-closed-output behavior in the future.

proppy commented on builder/internals.go:
```diff
@@ -433,7 +434,7 @@ func (b *Builder) pullImage(name string) (*imagepkg.Image, error) {
 	job.SetenvBool("parallel", true)
 	job.SetenvJson("authConfig", pullRegistryAuth)
-	job.Stdout.Add(b.OutOld)
+	job.Stdout.Add(ioutils.NopWriteCloser(b.OutOld))
```
wouldn't that mean that you also discard any output from pull?

jlhawn replied:
Sorry, I don't quite understand the question. :( The cause of the issue that you reported was that when the "pull" job ends it would close its Stdout which is also the "build" job's stdout. This makes the the "pull" job's Stdout Close no longer propagate and close the "build" Stdout.

jlhawn replied:
Oh wait - maybe you're confused by the name `NopWriteCloser`. This makes only the `Close` call be a nop - the writes still work and are written to the underlying writer.

LK4D4 commented:
LGTM
